### PR TITLE
GEOMESA-3311 Add subsampling to GraduatedQueryGuard

### PR DESCRIPTION
**Pull request by aheyne**
- Configurable subsample percentage based on size
- Configurable subsample attribute name
- Mix duration and percentage/attribute guard modes

### CODE REVIEW

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala [hunk 1]
```diff
@@ -110,17 +170,29 @@ object GraduatedQueryGuard extends LazyLogging {
...
+      } else None
+
+      val percentageAttr: Option[String] = if (limitsConfig.hasPath("attribute")) {
+        Some(limitsConfig.getString("attribute"))
+      } else None
+
+      new SizeAndLimits(size, duration, percentage, percentageAttr)
     }
-    evaluateLimits(confs.toSeq)
+    evaluateLimits(confs, sft)
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:196]: looks like you need to add back in the .toSeq to fix the 2.13 build

File: docs/user/datastores/query_interceptor.rst [hunk 2]
```diff
@@ -142,5 +147,59 @@ hours, or minutes. For example:
...
+
+.. code-block:: none
+
+    geomesa {
+      guard {
+        graduated {
+          "sftName" = [
+            { size = 1,  percentage = .8, attribute = "name" }
+            { size = 10, percentage = .5, attribute = "name"  }
+            { percentage = .1, attribute = "name" }
```
Comment (elahrvivaz) [inline docs/user/datastores/query_interceptor.rst:177]: maybe we could change these name to `sampling-percentage` and `sampling-attribute` or something similar so it's more descriptive? or alternatively have a nested block like `sampling: { percentage: .1, attribute: name}`

File: docs/user/datastores/query_interceptor.rst [hunk 3]
```diff
@@ -142,5 +147,59 @@ hours, or minutes. For example:
...
+            { duration = "1 day", percentage = .1, attribute = "name" }
+          ]
+        }
+      }
+    }
+
+In the above example, Any query with area less than 1 square degree will return all results and allow for any time
+range (given the time range doesn't trigger a full table scan). A query larger than 1 but less than 10 will block any
+query longer than 3 days AND only 50% of the results will be returned. Finally, any query larger than 10 must be less
+than 1 day in length AND only 10% of the results for any "name" will be returned.
```
Comment (elahrvivaz) [inline docs/user/datastores/query_interceptor.rst:202]: i find this behavior a little confusing but i think that's just how the graduated guard worked so no need to change it. thanks for the detailed explanation, it helps a lot!

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala [hunk 4]
```diff
@@ -32,12 +35,25 @@ class GraduatedQueryGuard extends QueryInterceptor with LazyLogging {
...
-    guardLimits = buildLimits(ConfigFactory.load().getConfigList(s"$ConfigPath.${sft.getTypeName}"))
+    guardLimits = buildLimits(ConfigFactory.load().getConfigList(s"$ConfigPath.${sft.getTypeName}"), sft)
     // this should be ensured during the loading of limits, but just double check so we can use `.last` safely below
     require(guardLimits.nonEmpty)
   }
 
-  override def rewrite(query: Query): Unit = {}
+  override def rewrite(query: Query): Unit = {
+    if (!disabled) {
+      val bounds = query.getFilter.accept(ExtractBoundsFilterVisitor.BOUNDS_VISITOR, null).asInstanceOf[Envelope]
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:45]: do you know if this operates on the default geom by default, or maybe we need to pass that in? technically there could be indices on other geometries, but i guess we don't know what index is being used yet at this point in the processing pipeline... maybe just using "any" geom would be best here (not sure if that's what it's currently doing or not)

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/TestGeoMesaDataStore.scala [hunk 5]
```diff
@@ -9,6 +9,7 @@
 package org.locationtech.geomesa.index
 
 import org.geotools.data.Query
+import org.geotools.util.factory.Hints
```
Comment (elahrvivaz) [inline geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/TestGeoMesaDataStore.scala:12]: unused import?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala (hunk 4 above)
Comment (aheyne) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:45]: The geotools docs say 
> all the literal geometry instances involved if spatial operations - using geom.getEnvelopeInternal().
which I would interpret to mean it considers all geometries included in the query.

**Review by elahrvivaz — APPROVED** [review]

**Pull request merged**